Handle invalid unicode while logging.

Paul-Ferrell opened the pull request:
Sometimes build processes will emit invalid unicode. In my case,
I ran into while building libtheora. (See merge request for traceback.)

This simply catches those errors and continues processes the log. I
normally wouldn't catch and release like this, but:

 1. This problem is obviously not endemic
 2. The alternative is to read in binary mode,
    and then decode while ignoring errors. If that's the solution that's
    prefered, it shouldn't be too big of a deal to tackle it.

One more thing
#############

An error like this might cascade; One invalid read puts the cursor at a bad position for reading the next unicode bytes, which cause the error again, and so on. With UTF8, this is pretty unlikely unless the text is very unicode heavy, as most characters should be readable as a single byte. 

```
==> Error: Exception occurred in writer daemon!
Traceback (most recent call last):
  File "/yellow/users/pflarr/repos/spack/lib/spack/llnl/util/tty/log.py", line 768, in _writer_daemon
    line = _retry(in_pipe.readline)()
  File "/yellow/users/pflarr/repos/spack/lib/spack/llnl/util/tty/log.py", line 830, in wrapped
    return function(*args, **kwargs)
  File "/var/lib/perceus/vnfs/asc-fe/rootfs/usr/lib64/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x97 in position 145: invalid start byte
```

Paul-Ferrell commented:
I should add that this solution will cause the log to lose the entire 'read' operation worth of log text, and not just the bad bytes (unless the read operation _is_ just the bad bytes).

becker33 commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -765,7 +765,13 @@ def _writer_daemon(stdin_multiprocess_fd, read_multiprocess_fd, write_fd, echo,
+                    # Sometimes build processes produce invalid unicode that 
+                    # the decoder can't handle. Ignore it and move on.
+                    except UnicodeDecodeError:
+                        continue
```
We should log the error here -- if we're going to lose part of the output, we should at least let the user know that we lost it.

bernhardkaindl replied:
This has been updated by Paul accordingly as far as I saw during rebase to current develop (it's one of the commits)

Paul-Ferrell commented:
How about just setting `line = '<data lost due to invalid unicode>\n'`.


________________________________
From: Greg Becker <notifications@github.com>
Sent: Tuesday, February 2, 2021 10:44 AM
To: spack/spack
Cc: Ferrell, Paul Steven; Author
Subject: [EXTERNAL] Re: [spack/spack] Handle invalid unicode while logging. (#21447)


@becker33 requested changes on this pull request.

________________________________

In lib/spack/llnl/util/tty/log.py<https://github.com/spack/spack/pull/21447#discussion_r568803391>:

> @@ -765,7 +765,13 @@ def _writer_daemon(stdin_multiprocess_fd, read_multiprocess_fd, write_fd, echo,

                 if in_pipe in rlist:
                     # Handle output from the calling process.
-                    line = _retry(in_pipe.readline)()
+                    try:
+                        line = _retry(in_pipe.readline)()
+                    # Sometimes build processes produce invalid unicode that
+                    # the decoder can't handle. Ignore it and move on.
+                    except UnicodeDecodeError:
+                        continue


We should log the error here -- if we're going to lose part of the output, we should at least let the user know that we lost it.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/spack/spack/pull/21447#pullrequestreview-581609606>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AMK6D5FXUOCVOFECL3PELT3S5A2XVANCNFSM4W7B4VAA>.

becker33 commented:
@Paul-Ferrell yeah that sounds about right.

Paul-Ferrell commented:
I'm not going to be able to get the code coverage tests to improve. It's a tiny change, and it's not something that can be universally tested for. On my linux with the python it comes with, I can force invalid unicode out stdout and force the error. That doesn't seem to be the case universally, however, so that code path doesn't always get executed. 

We continue to run into this problem as well when trying to build software for production usage.

Paul-Ferrell commented:
duplicates #15206

I like that solution better.

Paul-Ferrell commented:
Changed my mind, given that the other pull request has been sitting there for a year, and is sitting in decision paralysis on how to set things properly across the board.

becker33 commented:
@Paul-Ferrell sorry I fell down on reviewing this. It looks good to me know, but vermin is complaining that the changes to tty.log aren't python 2.6 compatible. I haven't quite figured out why yet.

becker33 commented:
Also, can you rebase this on the current Spack? That will cause it to run the newer tests that have been added in the meantime

bernhardkaindl commented:
@Paul-Ferrell I hope you don't mind: I rebased your commits to the current develop branch:

While I #23379 was intented to work with python2.6, it is obsolete because #27256 removed support for Python 2.6.

This PR works fine with 2.7, the package.py from my update of `gpgme` in #27604 is a reliable test case.

Also, my simple rebase of #23379 doesn't work, it hangs early: 
https://github.com/bernhardkaindl/spack/tree/bugfix/utf8-support-before-py-37
It gets stuck in the middle of the `configure` of `gpgme`, long before even the first non-ascii output of `bin/spack install -v --test=root gpgme`, which is in the output of the testsuite somewhere.

This new push is prepending `apply drop of Python 2.6 tests from PR#27256` to the stack of commits:

bernhardkaindl commented:
Ah, `vermin` doesn't look bad at all, it says this branch would now work with python 2.7, which is all we need since #27256 is removing support for 2.6. Thus, when it is merged, the python version obstacle might be solved:
```
Run vermin --backport argparse --violations --backport typing -t=2.6- -t=3.5- -vvv lib/spack/spack/ lib/spack/llnl/ bin/
Detecting python files..
Analyzing 483 files using 2 processes..
2.7, 3.0     /home/runner/work/spack/spack/lib/spack/spack/test/llnl/util/tty/log.py
  L65: 'sys.version_info.major' member requires 2.7, 3.0

Minimum required versions: 2.7, 3.5
Target versions not met:   2.6-, 3.5-
```

bernhardkaindl commented:
A reliabe testcase is #27604 when it's `var/spack/repos/builtin/packages/gpgme/package.py` is built using:
`LC_ALL=C bin/spack install -v --test=root gpgme@1.16.0`

bernhardkaindl commented:
Rebased to current develop. In a previous push, even the code coverage tests did pass. Luck?

haampie commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -62,6 +62,23 @@ def test_log_python_output_without_echo(capfd, tmpdir):
         assert capfd.readouterr()[0] == ''
 
 
+@pytest.mark.skipif(sys.version_info.major == 2 or sys.version_info.minor == 5,
```
```suggestion
@pytest.mark.skipif(sys.version_info.major == 2 or sys.version_info[:2] == (3, 5),
```

I agree your guard works and likely there's not gonna be a python 4.5, but ... you never know :p

Paul-Ferrell committed:
Handle byte sequences which are not encoded as UTF8 while logging.

Fix builds which produce a lines with non-UTF8 output while logging
The alternative is to read in binary mode, and then decode while
ignoring errors.

Co-authored-by: Bernhard Kaindl <bernhardkaindl7@gmail.com>
Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

bernhardkaindl commented:
The codecov coverage reduction is bogus, the line which it says is not covered can't have anything to do with the changes made.

https://codecov.io/gh/spack/spack/compare/6e095a9741dae0c08715fd2cf152558434413c4e...f7d43243db7f73bfeb227743ad65082dbc0f25a4/changes#

It seems to be allowed to merge it nonetheless. Waiting for some time, but for me it looks we can merge now.

The line which codecov says is no covered line is in `_find_provider` of `lib/spack/spack/spec.py`:
```py
2791             # Remove duplicate providers that can concretize to the same        
2792             # result.                                                           
2793             for provider in providers:                                          
2794                 for spec in providers:                                          
2795                     if spec is not provider and provider.satisfies(spec):       
2796 =>                      providers.remove(spec)                                                                                                                                      
```

haampie commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -62,6 +63,23 @@ def test_log_python_output_without_echo(capfd, tmpdir):
+            sys.stdout.buffer.write(b'\xc3\x28\n')
+
+        if sys.version_info.major == 2:
+            with open('foo.txt') as f:
+                assert(b'\xc3(\n')
```
```suggestion
            with open('foo.txt', 'rb') as f:
                assert(b'\xc3(\n') in f.read()
```

I think you forgot to read here?

bernhardkaindl replied:
Thanks, applied.

bernhardkaindl commented:
Thanks!

haampie commented:
Should this be backported without the test (which seems to require python 2.7)?

bernhardkaindl commented:
vermin told (I saw the vermin error in the github checks as well) that it's not compatible with python2.6, where is the comment from becker33 about it:
> @Paul-Ferrell sorry I fell down on reviewing this. It looks good to me know, but vermin is complaining that the changes to tty.log aren't python 2.6 compatible. I haven't quite figured out why yet.

Very few packages have such non-UTF8 characters in their logs left, so I think it's not of major importance to backport this, and it's not a new issue at all. It always existed.

I red that, using `io.open()` (I checked it appears to work, but it could have more overhead as it's not native code) would use backported python3 code in python2, which would make this compatible with python2.6 versions which have that backport but I didn't investigate that further.
